### PR TITLE
factor out layer copying into its own function

### PR DESCRIPTION
**Pull request by maxfreu**
Hi! I have factored out the layer copying from writelayers into its own function. So basically I changed (pseudocode):

```julia
function writelayers()
  create() do
    copystuff
  end
end
```

to 
```julia
function writelayers()
  create() do
    copylayers(src, dst)
  end
end

function copylayers(src, dst, ...)
  copystuff
end
```
That should now allow to copy data from and to arbitrary datasets, whether on disk or not. You can also give the layer indices you want to be copied. I added a tiny test for that. Furthermore I adjusted the docstrings for nicer rendering in the REPL. I'm not sure with the naming `copylayers`, is it ok? What do you think in general? Lastly, this supersedes https://github.com/yeesian/ArchGDAL.jl/pull/381, but I can take that part out if you want to merge this before the change to GDAL.

### CODE REVIEW

File: src/dataset.jl [hunk 1]
```diff
@@ -255,8 +257,36 @@ function writelayers(
...
+"""
+    copylayers(source_dataset, target_dataset; kwargs...)
+
+Copies a (sub)set of layers from source to target dataset. The keyword arguments are the same as for `writelayers`.
+Note that eventual layer options have to conform with the target dataset driver.
+"""
+function copylayers!(
+    source_dataset::AbstractDataset,
+    target_dataset::AbstractDataset;
+    layers = 0:(nlayer(source_dataset)-1),
```
Comment (yeesian) [inline src/dataset.jl:288]: Can we rename it from `layers` to `source_layer_indices` (or `source_layer_index` if it's `copylayer!`)?

File: src/dataset.jl [hunk 2]
```diff
@@ -255,8 +257,36 @@ function writelayers(
...
+    return nothing
+end
+
+"""
+    copylayers(source_dataset, target_dataset; kwargs...)
+
+Copies a (sub)set of layers from source to target dataset. The keyword arguments are the same as for `writelayers`.
+Note that eventual layer options have to conform with the target dataset driver.
+"""
+function copylayers!(
```
Comment (yeesian) [inline src/dataset.jl:285]: In the spirit of factoring things out -- should we make this `copylayer!(source_dataset, target_dataset, source_layer_index)` and letting the user write a loop over the layers they want to copy?

File: src/dataset.jl [hunk 3]
```diff
@@ -255,8 +257,36 @@ function writelayers(
...
-    options = [""],
-    layer_options = [""],
+    options = String[],
+    layers = 0:(nlayer(dataset)-1),
+    layer_options = String[],
+    chunksize::Integer = 20000,
+    use_gdal_copy::Bool = true,
+)
+    create(filename; driver = driver, options = options) do target
+        return copylayers!(
```
Comment (yeesian) [inline src/dataset.jl:267]: Can we add unit test(s) for `copylayer(s)!` without depending on it being called through `writelayers` in the unit tests?

**Review by yeesian — CHANGES_REQUESTED** [review]
Overall LGTM (modulo the changes requested), thank you!

File: src/dataset.jl (hunk 2 above)
Comment (maxfreu) [reply src/dataset.jl:299]: I think that would not give the user any benefit, so I'd keep it here.

**Comment by maxfreu** [comment]
I've just made the requested changes except for one :)

**Review by yeesian — APPROVED** [review]
@visr can I get your review here since it supersedes https://github.com/yeesian/ArchGDAL.jl/pull/381?

**Review by visr — APPROVED** [review]
Looks good to me, thanks!

**Comment by maxfreu** [comment]
Ok, done - sorry for the forth and back!

**Review by visr — APPROVED** [review]

**Pull request merged**